lighthouse/src/commonMain/kotlin/com/ivanempire/lighthouse/socket/BaseSocketListener.kt

```diff
@@ -0,0 +1,102 @@
+package com.ivanempire.lighthouse.socket
+
+import com.ivanempire.lighthouse.LighthouseLogger
+import com.ivanempire.lighthouse.models.Constants.DEFAULT_MULTICAST_ADDRESS
+import com.ivanempire.lighthouse.models.search.SearchRequest
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.isActive
+import java.net.DatagramPacket
+import java.net.InetAddress
+import java.net.InetSocketAddress
+import java.net.MulticastSocket
+
+/** Specific implementation of [SocketListener] */
+internal abstract class BaseSocketListener(
+    private val retryCount: Int,
+    private val logger: LighthouseLogger? = null,
+) : SocketListener {
+
+    private val multicastGroup: InetAddress by lazy {
+        InetAddress.getByName(DEFAULT_MULTICAST_ADDRESS)
+    }
+
+    protected abstract fun acquireMulticastLock()
+
+    private fun setupSocket(): MulticastSocket {
+        acquireMulticastLock()
+
+        val multicastSocket = MulticastSocket(null)
+        multicastSocket.reuseAddress = true
+        multicastSocket.broadcast = true
+        multicastSocket.loopbackMode = true // disable LoopbackMode
+
+        try {
+            multicastSocket.joinGroup(multicastGroup)
+            multicastSocket.bind(InetSocketAddress(MULTICAST_PORT))
+            logger?.logStatusMessage(TAG, "MulticastSocket has been setup")
+        } catch (ex: Exception) {
+            logger?.logErrorMessage(
+                TAG,
+                "Could finish setting up the multicast socket and group",
+                ex
+            )
+        }
+
+        return multicastSocket
+    }
+
+    override fun listenForPackets(searchRequest: SearchRequest): Flow<DatagramPacket> {
+        logger?.logStatusMessage(TAG, "Setting up datagram packet flow")
+        val multicastSocket = setupSocket()
+
+        return flow {
+                val datagramPacketRequest = searchRequest.toDatagramPacket(multicastGroup)
+
+                repeat(retryCount + 1) {
+                    if (!multicastSocket.isClosed) {
```

Comment:
This is to fix a crash I experienced. Not sure why the socket wasn't open, but there are apparently times when it's not.

Reply:
I am guessing you've seen this on desktop? Or did you observe this on Android?